uwblueprint/focus-on-nature
# fix: update user roles

## Notion ticket link

Task 2 for [Update users - role fix](https://www.notion.so/uwblueprintexecs/Update-users-role-fix-4c87f9e742cf4bcebd445d8a4af24f9d)



## Implementation description
* Added extra logic to add/remove the discriminator key `__t: "CampCoordinator"` depending on whether the user is an `Admin` or `CampCoordinator`
* Refactored the overall function a little bit



## Steps to test
1. Submit a PUT request to http://localhost:5000/users/:userId/. Choose a user in the database with a valid authId (you can use the user with ID `6209b327c1e74d00a5c0050a`) to pass into the `userId` field under "Path Variables". Can use this sample data (includes the fields for a `CampCoordinator`)
```
{
    "firstName": "ABC",
    "lastName": "DEF",
    "email": "hi@ok.com",
    "role": "CampCoordinator",
    "active": true,
    "campSessions": [
        "6229a30c827c860059d14407",
        "6229a3b684e22a00796f5c30"
    ]
}
```
2. Test changing between the `Admin` and `CampCoordinator` roles (if changing from `Admin` to `CampCoordinator`, then the discriminator key `__t` should be added and if changing from `CampCoordinator` to `Admin` then the discriminator key `__t` and `campSessions` fields should be removed)
3. Can test omitting certain fields in the update request (should still work)
4. Error cases: updating a non-existent user, non-existent role passed, any invalid primitive fields passed

## Notes
* Refer to https://github.com/uwblueprint/focus-on-nature/pull/18 and https://github.com/uwblueprint/focus-on-nature/pull/38 if you need more context on this function's implementation


## What should reviewers focus on?
* Any way to simplify this implementation?
* Any test cases missed?


## Checklist
